src/setup_runs/config_read_functions.py

```
@@ -0,0 +1,311 @@
+
+
+def parse_boolean_keys(config: dict,
+                       truevals: list[str] = ['true', '1', 't', 'y', 'yes'],
+                       falsevals: list[str] = ['false', '0', 'f', 'n', 'no'],
+                       bool_keys: list[str] = ["run_as_one_job", "submit_wrf_now", "submit_wps_component",
+                                               "only_edit_namelists",
+                                               "restart",
+                                               'delete_metem_files', "use_high_res_sst_data",
+                                               'regional_subset_of_grib_data']
+                       ) :
+    boolvals = truevals + falsevals
+
+    for bool_key in bool_keys :
+        assert (
+                config[bool_key].lower() in boolvals
+        ), f'Key {bool_key} not a recognised boolean value'
+        config[bool_key] = (config[bool_key].lower() in truevals)
+
+    return config
+
+
+## function to parse times
+def process_date_string(datestring) :
+    datestring = datestring.strip().rstrip()
+
+    ## get the timezone
+    if len(datestring) <= 19 :
+        tz = pytz.UTC
+        date = datetime.datetime.strptime(datestring, '%Y-%m-%d %H:%M:%S')
+    else :
+        tzstr = datestring[20 :]
+        tz = pytz.timezone(tzstr)
+        date = datetime.datetime.strptime(datestring, '%Y-%m-%d %H:%M:%S %Z')
+
+    date = tz.localize(date)
+
+    return date
+
+
+def load_wrf_config(filename: str) -> WRFConfig :
+    input_str = read_config_file(filename)
+
+    config = parse_config(input_str)
+
+    config = add_environment_variables(config=config, environmental_variables=os.environ)
+
+    config, iterationCount = substitute_variables(config)
+
+    # Iteration count for filling variables
+    assert iterationCount < 10, "Config key substitution exceeded iteration limit..."
+
+    ## parse boolean keys
+    config = parse_boolean_keys(config)
+
+    return WRFConfig(
+        project_root=config['project_root'],
```

This could be reduced to 

```
WRFConfig(**config)
```
then if any keys are missing it should raise an exception

---

Good to know. Took me a while to type all this.